sql/2024/cookies/0_create_desktop_cookies.sql

```diff
@@ -0,0 +1,39 @@
+-- Create an intermediate table containing all cookies that were set during the
+-- <DATE> crawl on <CLIENT> when visiting sites of rank <= <RANK>. This table
+-- can then be reused in consecutive queries without having to reextract the
+-- data every time
+-- Export the table as httparchive.almanac.DATE_CLIENT_RANK_cookies
+
+WITH intermediate_cookie AS (
+  SELECT
+    page,
+    rank,
+    JSON_VALUE(summary, '$.startedDateTime') AS startedDateTime,
+    cookie
+  FROM
+    `httparchive.all.pages`,
+    UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.cookies')) AS cookie
+  WHERE
+    date = '2024-06-01' AND
+    client = 'desktop' AND
+    rank <= 1000000
```

Comment:
Any reason to restrict it like this? Can we add a comment to explain why?

Reply:
This was to reduce the amount of data processed/parsed; without any cap on the rank, bigquery was returning an error "Response too large to return" or something like that. A solution is to specify a table where to output the results, but I don't have the permissions on the HTTP Archive project to create that. 

So, I created the table on a personal project/dataset, but then because I am on the free plan, there are some limits on amount of data I can store, and cookies on top 1M for both mobile and desktop were right below that limit, and top 1M seem plenty of enough data already.

Ideally, these cookies should not be in custom metrics but in their own separate table in my opinion. I believe I saw at some point an issue/discussion on HTTPArchive project where it was proposed to break down some fields in custom metrics in their table/column, instead of the current big blob.

Reply:
I can this table, which should be much cheaper to query:

```sql
-- Create an intermediate table containing all cookies that were set during the
-- <DATE> crawl on <CLIENT> when visiting sites of rank <= <RANK>. This table
-- can then be reused in consecutive queries without having to reextract the
-- data every time
-- Export the table as httparchive.almanac.DATE_CLIENT_RANK_cookies

CREATE TABLE `httparchive.almanac.cookies` 
(
  date DATE,
  client STRING,
  page STRING,
  root_page STRING,
  rank INTEGER,
  startedDateTime STRING,
  firstPartyCookie BOOL,
  name STRING,
  domain STRING,
  path STRING,
  expires STRING,
  size STRING,
  httpOnly STRING,
  secure STRING,
  session STRING,
  sameSite STRING,
  sameParty STRING,
  partitionKey STRING,
  partitionKeyOpaque STRING
)
PARTITION BY date
CLUSTER BY
  client, rank, page

AS

WITH intermediate_cookie AS (
  SELECT
    date,
    client,
    page,
    root_page,
    rank,
    JSON_VALUE(summary, '$.startedDateTime') AS startedDateTime,
    cookie
  FROM
    `httparchive.all.pages`,
    UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.cookies')) AS cookie
  WHERE
    date = '2024-06-01'
)

SELECT
  date,
  client,
  page,
  root_page,
  rank,
  startedDateTime,
  ENDS_WITH(NET.HOST(page), NET.REG_DOMAIN(JSON_VALUE(cookie, '$.domain'))) AS firstPartyCookie,
  JSON_VALUE(cookie, '$.name') AS name,
  JSON_VALUE(cookie, '$.domain') AS domain,
  JSON_VALUE(cookie, '$.path') AS path,
  JSON_VALUE(cookie, '$.expires') AS expires,
  JSON_VALUE(cookie, '$.size') AS size,
  JSON_VALUE(cookie, '$.httpOnly') AS httpOnly,
  JSON_VALUE(cookie, '$.secure') AS secure,
  JSON_VALUE(cookie, '$.session') AS session,
  JSON_VALUE(cookie, '$.sameSite') AS sameSite,
  JSON_VALUE(cookie, '$.sameParty') AS sameParty,
  JSON_VALUE(cookie, '$.partitionKey') AS partitionKey,
  JSON_VALUE(cookie, '$.partitionKeyOpaque') AS partitionKeyOpaque
FROM intermediate_cookie
```

Could you update this query to this:

```sql
INSERT INTO `httparchive.almanac.cookies` 
WITH intermediate_cookie AS (
  SELECT
    date,
...
```

 But no need to run as I've run it.

Then use this new `httparchive.almanac.cookies` table in your other queries? I presume it won't change your results much.

Reply:
Nice! Thank you! Looking into the new table you created and updating the queries for the cookies chapter accordingly.